File: hello-grpc-swift/Sources/Common/Connection.swift
import Foundation
import Logging

public protocol Connection {
    var host: String? { get }
    var port: Int? { get }
    var backendHost: String? { get }
    var backendPort: Int? { get }
    var hasBackend: Bool { get }
}

public class HelloConn: Connection {
    let logger = Logger(label: "HelloConn")
    public var host: String?
    public var backendHost: String?
    public var port: Int?
    public var backendPort: Int?

    public var hasBackend: Bool {
        backendHost != nil && !backendHost!.isEmpty
    }

    public init() {
        // Get backend configuration from environment
        if let envHost = ProcessInfo.processInfo.environment["GRPC_SERVER"], !envHost.isEmpty {
            host = envHost
        } else {
            // 检测是否为 docker 环境
            let dockerHost: String = {
                if let cgroup = try? String(contentsOfFile: "/proc/1/cgroup", encoding: .utf8),
                   cgroup.contains("docker") || cgroup.contains("containerd")
                {
                    return "0.0.0.0"
                }
                // 兼容 macOS 本地开发
                #if os(macOS)
                    return "127.0.0.1"
                #else
                    return "0.0.0.0"
                #endif
            }()
            host = dockerHost
            logger.info("GRPC_SERVER 未设置，自动使用默认 host: \(dockerHost)")
        }

        if let portStr = ProcessInfo.processInfo.environment["GRPC_SERVER_PORT"] {
            if let port0 = Int(portStr) {
                port = port0
            }
        } else {
            // Set default port when not specified in environment
            port = 9996
        }

        // Get backend configuration from environment
        backendHost = ProcessInfo.processInfo.environment["GRPC_HELLO_BACKEND"]

        if let backendPortStr = ProcessInfo.processInfo.environment["GRPC_HELLO_BACKEND_PORT"] {
            if let port = Int(backendPortStr) {
                backendPort = port
            }
        }

        logger.info("Hello service port: \(port ?? 9996)")
        if hasBackend {
            logger.info("Backend configured: \(backendHost!):\(backendPort ?? 9996)")
        }
    }
}

File: hello-grpc-php/conn/Connection.php
<?php

use Monolog\Logger;

/**
 * Connection configuration for the gRPC server and client
 * 
 * This class handles all connection-related configuration including:
 * - TLS certificate loading and validation
 * - Backend service discovery and connection management
 * - Environment variable configuration
 * 
 * Handles TLS certificate management, server/client port configuration,
 * and backend connection details.
 * 
 * @author Hello gRPC Team
 */
class Connection
{
    /** @var string Server port */
    public $port;
    
    /** @var bool Whether to use TLS */
    public $isSecure;
    
    /** @var string Backend host (for proxying requests) */
    public $backendHost;
    
    /** @var string Backend port (for proxying requests) */
    public $backendPort;
    
    /** @var string Path to server certificate */
    public $certPath;
    
    /** @var string Path to server private key */
    public $keyPath;
    
    /** @var string Path to server certificate chain */
    public $chainPath;
    
    /** @var string Path to root certificate */
    public $rootCertPath;
    
    /** @var \Psr\Log\LoggerInterface Logger instance */
    private $logger;

    /**
     * Constructor
     */
    public function __construct()
    {
        global $log;
        $this->logger = $log ?? new Logger('Connection');
        
        // Initialize from environment variables with defaults
        $this->port = getenv('GRPC_SERVER_PORT') ?: '9996';
        $this->isSecure = getenv('GRPC_HELLO_SECURE') === 'Y';
        
        // Check for backend configuration
        $this->backendHost = getenv('GRPC_HELLO_BACKEND');
        $this->backendPort = getenv('GRPC_HELLO_BACKEND_PORT');
        
        // Get certificate paths based on platform
        $this->setupCertificatePaths();
        
        $this->logger->info(sprintf("Connection configuration: port=%s, tls=%s", 
            $this->port, $this->isSecure ? 'enabled' : 'disabled'));
            
        if ($this->hasBackend()) {
            $this->logger->info(sprintf("Backend configuration: host=%s, port=%s", 
                $this->backendHost, $this->backendPort ?: $this->port));
        }
    }

    /**
     * Check if a backend service is configured
     * @return bool True if backend is configured
     */
    public function hasBackend(): bool
    {
        return !empty($this->backendHost);
    }
    
    /**
     * Setup certificate paths based on the current platform (Windows/macOS/Linux)
     */
    private function setupCertificatePaths(): void
    {
        // Use custom base path from environment if available
        $basePath = getenv('CERT_BASE_PATH');
        
        // Otherwise, use platform-specific default
        if (empty($basePath)) {
            $osName = PHP_OS_FAMILY;
            
            if ($osName === 'Windows') {
                $basePath = 'd:\\garden\\var\\hello_grpc\\server_certs';
            } elseif ($osName === 'Darwin') {
                // macOS uses a different path structure
                $basePath = '/var/hello_grpc/server_certs';
            } else {
                // Linux/Unix default
                $basePath = '/var/hello_grpc/server_certs';
            }
            
            $this->logger->info(sprintf("Using platform-specific (%s) certificate path: %s", 
                $osName, $basePath));
        } else {
            $this->logger->info(sprintf("Using custom certificate path from environment: %s", 
                $basePath));
        }
        
        // Define certificate paths
        $this->certPath = $basePath . DIRECTORY_SEPARATOR . 'cert.pem';
        $this->keyPath = $basePath . DIRECTORY_SEPARATOR . 'private.key';
        $this->chainPath = $basePath . DIRECTORY_SEPARATOR . 'full_chain.pem';
        $this->rootCertPath = $basePath . DIRECTORY_SEPARATOR . 'myssl_root.cer';
    }
    
    /**
     * Validate certificate files exist and are readable
     * @return bool True if certificates are valid, false otherwise
     */
    public function validateCertificates(): bool
    {
        if (!$this->isSecure) {
            return false;
        }
        
        // Check required certificate files
        $certFiles = [
            'Certificate' => $this->certPath,
            'Private key' => $this->keyPath
        ];
        
        foreach ($certFiles as $name => $path) {
            if (!file_exists($path)) {
                $this->logger->warning("$name file not found: $path");
                return false;
            }
            
            if (!is_readable($path)) {
                $this->logger->warning("$name file not readable: $path");
                return false;
            }
        }
        
        // Optional root certificate
        if (file_exists($this->rootCertPath)) {
            if (is_readable($this->rootCertPath)) {
                $this->logger->info("Root certificate found: " . $this->rootCertPath);
            } else {
                $this->logger->warning("Root certificate found but not readable: " . $this->rootCertPath);
            }
        } else {
            $this->logger->info("Root certificate not found, client verification disabled");
        }
        
        return true;
    }
}